File: backend/python/one_touch_loader/api/schemas/player_rankings.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


class PlayerRatingReferenceOut(BaseModel):
    start_season_name: str
    end_season_name: str
    minimum_rated_matches: int = Field(description="각 리그·시즌 중반 이후의 최소 평점 경기 수예요.")
    early_season_minimum_rated_matches: Literal[1] = 1
    competition_ids: list[int]
    sample_count: int
    updated_at: datetime


class PlayerRankingOut(BaseModel):
    rank: int = Field(description="반올림 전 평균 평점 순위. 평균도 같으면 공동 순위예요.")
    player_id: int
    player_name: str
    player_image: str | None
    rated_matches: int
    average_rating: float
    percentile_score: float = Field(description="저장된 백분위. 조회한 선수 목록으로 다시 환산하지 않아요.")
    display_score: float = Field(description="백분위를 고정 환산한 1Touch 점수예요. 상위 1%는 90점이며 환산 후 소수점 한 자리로 반올림해요.")
    updated_at: datetime


class PlayerRankingsResponse(BaseModel):
    competition_id: int
    season_id: int
    season_name: str
    method: Literal["cumulative_all_leagues_percentile"] = "cumulative_all_leagues_percentile"
    reference: PlayerRatingReferenceOut
    total: int
    limit: int
    offset: int
    items: list[PlayerRankingOut]

File: frontend/lib/models/player_rankings.dart
import 'package:flutter/foundation.dart';

enum PlayerRankingMethod { fixedHistoricalPercentile }

@immutable
class PlayerRatingReference {
  const PlayerRatingReference({
    required this.startSeasonName,
    required this.endSeasonName,
    required this.minimumRatedMatches,
    required this.sampleCount,
    required this.frozenAt,
  });

  final String startSeasonName;
  final String endSeasonName;
  final int minimumRatedMatches;
  final int sampleCount;
  final DateTime frozenAt;
}

@immutable
class RankedPlayer {
  const RankedPlayer({
    required this.rank,
    required this.playerId,
    required this.playerName,
    required this.playerImage,
    required this.ratedMatches,
    required this.averageRating,
    required this.percentileScore,
    required this.displayScore,
    required this.updatedAt,
  });

  final int rank;
  final int playerId;
  final String playerName;
  final String? playerImage;
  final int ratedMatches;
  final double averageRating;
  final double percentileScore;
  final double displayScore;
  final DateTime updatedAt;
}

@immutable
class PlayerRankingsPage {
  PlayerRankingsPage({
    required this.competitionId,
    required this.seasonId,
    required this.seasonName,
    required this.method,
    required this.reference,
    required this.total,
    required this.limit,
    required this.offset,
    required List<RankedPlayer> items,
  }) : items = List.unmodifiable(items);

  final int competitionId;
  final int seasonId;
  final String seasonName;
  final PlayerRankingMethod method;
  final PlayerRatingReference reference;
  final int total;
  final int limit;
  final int offset;
  final List<RankedPlayer> items;
}
